### PR TITLE
geph.{cli,gui}: 4.7.8 -> 4.10.1

### PR DESCRIPTION
**Pull request by penalty1083**
###### Description of changes

Closes https://github.com/NixOS/nixpkgs/pull/240233

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by lilyinstarlight — COMMENTED** [review]
Could you update this to 4.8.9, and also can you just vendor only the Cargo.lock for the GUI rather than both?

It seems unnecessary for the CLI and we really want to avoid copying in large files to nixpkgs when not necessary :)

Thank you!

**Comment by penalty1083** [comment]
@lilyinstarlight Sorry for the late reply. Suggestions are applied.

File: pkgs/applications/networking/geph/default.nix
Comment (lilyinstarlight) [inline pkgs/applications/networking/geph/default.nix:90]: Is `pnpm` reproducible like this? I imagine it's stable for a given version, but I would be interested to know if this output is stable across pnpm versions

If I find time I'll do some testing myself, but I'm admittedly not as familiar with pnpm and I don't know if they guarantee a particular store layout that's byte-for-byte reproducible (after removing timestamps and sorting keys that is, as this implementation seems to do)

Comment (penalty1083) [reply pkgs/applications/networking/geph/default.nix:90]: I don't know if it's stable across pnpm versions. The same trick has been used in [pot](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/pot/default.nix) and looks like it works well. Since the store is versioned, I would expect that the format won't be changed if the store version is not changed. I'm not familiar with pnpm and this trick is borrowed from pot. I thought if the store format is changed in the future we need to pin the pnpm version, as what we do for other packages.

Comment (lilyinstarlight) [reply pkgs/applications/networking/geph/default.nix:90]: Hmmm, I'll try to look at the pnpm documentation and do testing to confirm whether that is the case in the next week or so

Either way though, until we have an official fetcher though it might be best to avoid these custom fetchers if they aren't vetted

Comment (penalty1083) [reply pkgs/applications/networking/geph/default.nix:90]: @lilyinstarlight Kindly ping :). Instead of waiting for a "proper" solution forever, I thought it's better to merge this to unblock the update of this security-essential program.

This workground works well for a long time and I bet it will work much better than a home-made, broken by design "proper" fetcher. And I don't understand why a home-made fetcher is better. If pnpm decides to change the format of the local store, then no matter how we fetch the data, we need to change our code. Instead, we can pin the pnpm version so that the format won't be changed.

Comment (lilyinstarlight) [reply pkgs/applications/networking/geph/default.nix:90]: As for unblocking this PR, would you be open to just switching it back to the package-lock.json approach for now and work on the pnpm fetcher in a new PR?

I do appreciate your work on this!

Comment (93221) [reply pkgs/applications/networking/geph/default.nix:90]: > We've had issues before where even things as dumb as structured parsers (e.g. jq or yq) subtly changed their output and broke any FOD fetchers that were transforming data with them. So when we create fetchers that always output a stable format, we don't have to care when pnpm or jq or anything updates

Good point! If you are worrying about jq changing the output format we can minimize the json after sorting the items.

> Pinning versions sorta helps, but then we're having to maintain old versions for not much reason when we can have purpose-built code where versions no longer matter (edit: to be clear, we do sometimes use original tooling when creating FODs, like with cargo, but those projects also generally explicitly promise a future-proof reproducible output, which I can't tell is the case with pnpm)

It's not easier to maintain a custom fetcher though. It's not enough to output a stable FOD, it must also work. But a home-made fetcher usually doesn't work in some cases. It's hard to write a correct fetcher.

> As for unblocking this PR, would you be open to just switching it back to the package-lock.json approach for now and work on the pnpm fetcher in a new PR?

It should work, yes. But It's generally against the principle that we should avoid adding large lock files into this repo. And it means I need to use different packages from upstream.

Comment (lilyinstarlight) [reply pkgs/applications/networking/geph/default.nix:90]: > Good point! If you are worrying about jq changing the output format we can minimize the json after sorting the items.

Jq's compact output should be fine there, I don't expect that to change

> It's not easier to maintain a custom fetcher though. It's not enough to output a stable FOD, it must also work. But a home-made fetcher usually doesn't work in some cases. It's hard to write a fetcher.

If they have trouble working with the upstream tooling (e.g. in some of the npm problems we've had), it's almost universally because the upstream tooling couldn't work reproducibly in that case anyway

I've unfortunately been personally bitten enough times by custom and complex FODs in derivations like this one that I've become wary :/

> It should work, yes. But It's generally against the principle that we should avoid adding large lock files into this repo. And it means I need to use different packages from upstream.

If you're going to be this against changing it, I can still review this as-is and don't mind much. It's just going to take a bit longer to do so

I'll try to carve out extra time in the next few days to verify FOD reproducibility here (and for pot, which seemed to have been merged without verification...), since unlike input-addressed derivations, FODs have a hard requirement on byte-for-byte reproducibility

**Comment by Artturin** [comment]
@ofborg build geph.cli geph.gui

FYI in the future you can use `geph.{cli,gui}: ...` in the commit title to make ofborg build the non-nonfree one automatically.
https://github.com/NixOS/ofborg#automatic-building

ofborg just got support for brace expansion so i gotta advertise it :)

**Review by lilyinstarlight — APPROVED** [review]
I don't have time to adequately getting around to verifying this

Because both that geph currently does not build/work and that already-merged packages (e.g. pot) will need to be evaluated too to determine if this FOD is stable, there's no reason to hold this up over it and I will just come back to this derivation when I come back to pot and others that use this FOD hack

I'll approve this change for now and will probably open a new issue about investigating these FODs later

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
seems we're up to version 4.9.6 as per https://github.com/geph-official/geph4-client/tags, pr title should also be updated accordingly

**Comment by penalty1083** [comment]
Sorry for the late reply. I updated geph to 4.9.9.

**Comment by penalty1083** [comment]
Updated to 4.10.1.

**Comment by penalty1083** [comment]
Thanks!

**Pull request merged**
